include/llama/Copy.hpp

```diff
@@ -0,0 +1,381 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#pragma once
+
+#include "View.hpp"
+#include "mapping/AoSoA.hpp"
+#include "mapping/SoA.hpp"
+
+#include <cstring>
+#include <numeric>
+
+namespace llama
+{
+    namespace internal
+    {
+        inline void parallel_memcpy(
+            std::byte* dst,
+            const std::byte* src,
+            std::size_t size,
+            std::size_t threadId = 0,
+            std::size_t threadCount = 1,
+            decltype(std::memcpy) singleThreadMemcpy = std::memcpy)
+        {
+            const auto sizePerThread = size / threadCount;
+            const auto sizeLastThread = sizePerThread + size % threadCount;
+            const auto sizeThisThread = threadId == threadCount - 1 ? sizeLastThread : sizePerThread;
+            singleThreadMemcpy(dst + threadId * sizePerThread, src + threadId * sizePerThread, sizeThisThread);
+        }
+    } // namespace internal
+
+    /// Direct memcpy from source view blobs to destination view blobs. Both views need to have the same mappings with
+    /// the same array dimensions.
+    /// @param threadId Optional. Zero-based id of calling thread for multi-threaded invocations.
+    /// @param threadCount Optional. Thread count in case of multi-threaded invocation.
+    template <typename Mapping, typename SrcBlob, typename DstBlob>
+    void blobMemcpy(
+        const View<Mapping, SrcBlob>& srcView,
+        View<Mapping, DstBlob>& dstView,
+        std::size_t threadId = 0,
+        std::size_t threadCount = 1)
+    {
+        // TODO: we do not verify if the mappings have other runtime state than the array dimensions
+        if (srcView.mapping.arrayDims() != dstView.mapping.arrayDims())
+            throw std::runtime_error{"Array dimensions sizes are different"};
+
+        // TODO: this is maybe not the best parallel copying strategy
+        for (std::size_t i = 0; i < Mapping::blobCount; i++)
+            internal::parallel_memcpy(
+                &dstView.storageBlobs[i][0],
+                &srcView.storageBlobs[i][0],
+                dstView.mapping.blobSize(i),
+                threadId,
+                threadCount);
+    }
+
+    /// Field-wise copy from source to destination view. Both views need to have the same array and record dimensions.
+    /// @param threadId Optional. Thread id in case of multi-threaded copy.
+    /// @param threadCount Optional. Thread count in case of multi-threaded copy.
+    template <typename SrcMapping, typename SrcBlob, typename DstMapping, typename DstBlob>
+    void fieldWiseCopy(
+        const View<SrcMapping, SrcBlob>& srcView,
+        View<DstMapping, DstBlob>& dstView,
+        std::size_t threadId = 0,
+        std::size_t threadCount = 1)
+    {
+        // TODO: think if we can remove this restriction
+        static_assert(std::is_same_v<typename SrcMapping::RecordDim, typename DstMapping::RecordDim>);
```

Comment:
Maybe add an error message here.

Reply:
How much value would such an error message be: "The source and destination record dimensions must be the same". Because this is exactly what is already written here. I wonder if I can convey any more meaningful help.

Reply:
Well, the message you just suggested is immediately human-readable. Without the message, the user possibly needs to parse a C++ template novel first.

Reply:
Which they need to parse anyway, to know which mapping exactly, and from which copy invocation the error was raised ;) But fine, I will add it.

Reply:
Added.